File: Raven.Client.Java/src/main/java/net/ravendb/client/connection/AdminRequestCreator.java
package net.ravendb.client.connection;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.closure.Function2;
import net.ravendb.abstractions.closure.Function3;
import net.ravendb.abstractions.data.DatabaseDocument;
import net.ravendb.abstractions.data.HttpMethods;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.client.connection.implementation.HttpJsonRequest;
import net.ravendb.client.extensions.MultiDatabase;
import net.ravendb.client.utils.UrlUtils;


public class AdminRequestCreator {
  //url, method
  private final Function2<String, HttpMethods, HttpJsonRequest> createRequestForSystemDatabase;
  private final Function2<String, HttpMethods, HttpJsonRequest> createRequest;

  //currentServerUrl, operationUrl, method
  private final Function3<String, String, HttpMethods, HttpJsonRequest> createReplicationAwareRequest;

  public AdminRequestCreator(Function2<String, HttpMethods, HttpJsonRequest> createRequestForSystemDatabase,
    Function2<String, HttpMethods, HttpJsonRequest> createRequest,
    Function3<String, String, HttpMethods, HttpJsonRequest> createReplicationAwareRequest) {
    super();
    this.createRequestForSystemDatabase = createRequestForSystemDatabase;
    this.createRequest = createRequest;
    this.createReplicationAwareRequest = createReplicationAwareRequest;
  }

  public HttpJsonRequest createDatabase(DatabaseDocument databaseDocument, Reference<RavenJObject> docRef) {
    if (!databaseDocument.getSettings().containsKey("Raven/DataDir")) {
      throw new IllegalStateException("The Raven/DataDir setting is mandatory");
    }
    String dbname = databaseDocument.getId().replace("Raven/Databases/", "");
    MultiDatabase.assertValidDatabaseName(dbname);
    RavenJObject doc = RavenJObject.fromObject(databaseDocument);
    doc.remove("id");
    docRef.value = doc;

    return createRequestForSystemDatabase.apply("/admin/databases/" + UrlUtils.escapeDataString(dbname), HttpMethods.PUT);
  }

  public HttpJsonRequest deleteDatabase(String databaseName, boolean hardDelete) {
    String deleteUrl = "/admin/databases/" + UrlUtils.escapeDataString(databaseName);

    if (hardDelete) {
      deleteUrl += "?hard-delete=true";
    }

    return createRequestForSystemDatabase.apply(deleteUrl, HttpMethods.DELETE);
  }

  public HttpJsonRequest stopIndexing(String serverUrl) {
    return createReplicationAwareRequest.apply(serverUrl, "/admin/StopIndexing", HttpMethods.POST);
  }

  public HttpJsonRequest startIndexing(String serverUrl) {
    return createReplicationAwareRequest.apply(serverUrl, "/admin/StartIndexing", HttpMethods.POST);
  }

  public HttpJsonRequest adminStats() {
    return createRequestForSystemDatabase.apply("/admin/stats", HttpMethods.GET);
  }

  public HttpJsonRequest startBackup(String backupLocation, DatabaseDocument databaseDocument, Reference<RavenJObject> backupSettingsRef) {
    RavenJObject backupSettings = new RavenJObject();
    backupSettingsRef.value = backupSettings;

    backupSettings.add("BackupLocation", backupLocation);
    backupSettings.add("DatabaseDocument", RavenJObject.fromObject(databaseDocument));

    return createRequest.apply("/admin/backup", HttpMethods.POST);
  }

  public HttpJsonRequest startRestore(String restoreLocation, String databaseLocation, String databaseName, boolean defrag, Reference<RavenJObject> restoreSettingsRef) {
    RavenJObject restoreSettings = new RavenJObject();
    restoreSettingsRef.value = restoreSettings;
    restoreSettings.add("RestoreLocation", restoreLocation);
    restoreSettings.add("DatabaseLocation", databaseLocation);
    restoreSettings.add("DatabaseName", databaseName);

    return createRequest.apply("/admin/restore?defrag=" + defrag, HttpMethods.POST);

  }

  public HttpJsonRequest indexingStatus(String serverUrl) {
    return createReplicationAwareRequest.apply(serverUrl, "/admin/IndexingStatus", HttpMethods.GET);
  }

  public HttpJsonRequest compactDatabase(String databaseName) {
    return createRequestForSystemDatabase.apply("/admin/compact?database=" + databaseName, HttpMethods.POST);
  }
}

File: Raven.Client.Lightweight/Connection/AdminRequestCreator.cs
// -----------------------------------------------------------------------
//  <copyright file="AdminDatabaseCommandsBase.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------
using System;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Raven.Abstractions.Data;
using Raven.Abstractions.Extensions;
using Raven.Client.Connection.Implementation;
using Raven.Client.Extensions;
using Raven.Json.Linq;

namespace Raven.Client.Connection
{
	public class AdminRequestCreator
	{
		// url, method
		private readonly Func<string, HttpMethod, HttpJsonRequest> createRequestForSystemDatabase;

		// currentServerUrl, operationUrl, method
		private readonly Func<string, string, HttpMethod, HttpJsonRequest> createReplicationAwareRequest;

		public AdminRequestCreator(Func<string, HttpMethod, HttpJsonRequest> createRequestForSystemDatabase, Func<string, string, HttpMethod, HttpJsonRequest> createReplicationAwareRequest)
		{
			this.createRequestForSystemDatabase = createRequestForSystemDatabase;
			this.createReplicationAwareRequest = createReplicationAwareRequest;
		}

		public HttpJsonRequest CreateDatabase(DatabaseDocument databaseDocument, out RavenJObject doc)
		{
			if (databaseDocument.Settings.ContainsKey("Raven/DataDir") == false)
				throw new InvalidOperationException("The Raven/DataDir setting is mandatory");
			var dbname = databaseDocument.Id.Replace("Raven/Databases/", "");
            MultiDatabase.AssertValidName(dbname);
			doc = RavenJObject.FromObject(databaseDocument);
			doc.Remove("Id");

			return createRequestForSystemDatabase("/admin/databases/" + Uri.EscapeDataString(dbname), HttpMethod.Put);
		}

		public HttpJsonRequest DeleteDatabase(string databaseName, bool hardDelete)
		{
			var deleteUrl = "/admin/databases/" + Uri.EscapeDataString(databaseName);

			if(hardDelete)
				deleteUrl += "?hard-delete=true";

			return createRequestForSystemDatabase(deleteUrl, HttpMethod.Delete);
		}

		public HttpJsonRequest StopIndexing(string serverUrl)
		{
			return createReplicationAwareRequest(serverUrl, "/admin/StopIndexing", HttpMethod.Post);
		}

        public HttpJsonRequest StartIndexing(string serverUrl, int? maxNumberOfParallelIndexTasks)
        {
            var url = "/admin/StartIndexing";
            if (maxNumberOfParallelIndexTasks.HasValue)
            {
                url += "?concurrency=" + maxNumberOfParallelIndexTasks.Value;
            }

			return createReplicationAwareRequest(serverUrl, url, HttpMethod.Post);
		}

		public HttpJsonRequest AdminStats()
		{
			return createRequestForSystemDatabase("/admin/stats", HttpMethod.Get);
		}

		public HttpJsonRequest StartBackup(string backupLocation, DatabaseDocument databaseDocument, string databaseName, bool incremental)
		{
            if (databaseName == Constants.SystemDatabase)
            {
				return createRequestForSystemDatabase("/admin/backup", HttpMethod.Post);
            }
			return createRequestForSystemDatabase("/databases/" + databaseName + "/admin/backup?incremental=" + incremental, HttpMethod.Post);
            
		}

		public HttpJsonRequest CreateRestoreRequest()
		{
			return createRequestForSystemDatabase("/admin/restore", HttpMethod.Post);
		}

		public HttpJsonRequest IndexingStatus(string serverUrl)
		{
			return createReplicationAwareRequest(serverUrl, "/admin/IndexingStatus", HttpMethod.Get);
		}

		public HttpJsonRequest CompactDatabase(string databaseName)
		{
			return createRequestForSystemDatabase("/admin/compact?database=" + databaseName, HttpMethod.Post);
		}

		public HttpJsonRequest GetDatabaseConfiguration(string serverUrl)
		{
			return createReplicationAwareRequest(serverUrl, "/debug/config", HttpMethod.Get);
		}

        /// <summary>
        /// Gets the list of databases from the server asynchronously
        /// </summary>
		public async Task<string[]> GetDatabaseNamesAsync(int pageSize, int start = 0, CancellationToken token = default (CancellationToken))
        {
	        using (var requestForSystemDatabase = createRequestForSystemDatabase(string.Format(CultureInfo.InvariantCulture, "/databases?pageSize={0}&start={1}", pageSize, start), HttpMethod.Get))
	        {
				var result = await requestForSystemDatabase.ReadResponseJsonAsync().WithCancellation(token).ConfigureAwait(false);
				var json = (RavenJArray)result;
				return json.Select(x => x.ToString())
					.ToArray();
	        }
        }
	}
}
